lib/github/ldap/forest_search.rb

```diff
@@ -0,0 +1,99 @@
+require 'github/ldap/instrumentation'
+
+module GitHub
+  class Ldap
+    # For ActiveDirectory environments that have a forest with multiple domain controllers,
+    # this strategy class allows for entry searches across all domains in that forest.
+    class ForestSearch
+      include Instrumentation
+
+      # Build a new GitHub::Ldap::ForestSearch instance
+      #
+      # connection:     GitHub::Ldap object representing the main AD connection.
+      # naming_context: The Distinguished Name (DN) of this forest's Configuration
+      #                 Naming Context, e.g., "CN=Configuration,DC=ad,DC=ghe,DC=com"
+      #
+      # See: https://technet.microsoft.com/en-us/library/aa998375(v=exchg.65).aspx
+      #
+      def initialize(connection, naming_context)
+        @naming_context = naming_context
+        @connection = connection
+      end
+
+      # Search over all domain controllers in the ActiveDirectory forest.
+      #
+      # options: is a hash with the same options that Net::LDAP::Connection#search supports.
+      # block: is an optional block to pass to the search.
+      #
+      # If no domain controllers are found in the forest, fall back on searching
+      # the main GitHub::Ldap object in @connection.
+      #
+      # If @forest is populated, iterate over each domain controller and perform
+      # the requested search, excluding domain controllers whose naming context
+      # is not in scope for the search base DN defined in options[:base].
+      #
+      def search(options, &block)
+        instrument "forest_search.github_ldap" do
+          if forest.empty?
+            @connection.search(options, &block)
+          else
+            forest.each_with_object([]) do |(ncname, connection), res|
+              if options[:base].end_with?(ncname)
```

Comment:
This needs to inspect the Domain Components to make sure the that right Domain Controller is performing the search.